### PR TITLE
Algorithm - Remove all-reduce for routing probabilities in global-batch load balancing loss

### PR DESCRIPTION
**Pull request by qghuxmu**
Remove all-reduce for `routing_prob` in global-batch load balancing loss.

The theory is that, when calculating gradient for `all_reduced(routing_prob) * all_reduced(routing_count)`:
- We don't need the gradient for `routing_count` part because it's not in model weights and not differentiable
- The gradient of `torch.distributed.all_reduce` node is IdentityOp

As a result, removing all-reduce for `routing_prob` will not affect the gradient of `routing_prob` part, and save latency.

### CODE REVIEW

**Comment by yzygitzh** [comment]
@qghuxmu Please describe the theory why we can remove all-reduce for probs in seq parallelism in the PR description.

File: megatron/core/transformer/moe/moe_utils.py [hunk 1]
```diff
@@ -154,9 +154,7 @@ def global_batch_load_balancing_loss_func(
     # sequence.
     if sequence_partition_group is not None:
         routing_prob /= sequence_partition_group.size()
-        routing_prob_count = torch.stack([routing_prob, routing_count])
-        torch.distributed.all_reduce(routing_prob_count, op=torch.distributed.ReduceOp.SUM, group=sequence_partition_group)
-        routing_prob, routing_count = routing_prob_count[0], routing_prob_count[1]
+        torch.distributed.all_reduce(routing_count, op=torch.distributed.ReduceOp.SUM, group=sequence_partition_group)
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/moe_utils.py:157]: Let's borrow the comment in switch lbl to make the logic more clear here:
```
# We can keep `aggregated_probs_per_expert` local since we don't need the gradient for
# `tokens_per_expert`, saving one allreduce operation for `aggregated_probs_per_expert`.
```

**Review by yzygitzh — APPROVED** [review]

**Comment by yzygitzh** [comment]
After offline discussion with @cp5555 , we decide to hold this revision, since it only produce correct results when `routing_count` is not differentiable, which is not a general case. Besides, the intra sequence parallelism group all-reduce is not costy.
